### PR TITLE
run tests in parallel by equally dividing them between workers

### PR DESCRIPTION
**Pull request by vladima**
difference from current implementation of `runtests-parallel`: instead of partitioning statically on per-category basis we first load a list of tests and equally divide every category based on the number of workers (by default is set to `require("os").cpus().length`).
Running times (with `lint=false`) on  Intel(R) Core(TM) i7 CPU 860  @ 2.80GHz, RAM 16 GB:

 - jake runtests lint=false - 5m45.681s
 - jake runtests-parallel lint=false - 1m42.367s

### CODE REVIEW

File: Jakefile.js [hunk 1]
```diff
@@ -717,45 +730,71 @@ function runConsoleTests(defaultReporter, defaultSubsets) {
...
-            complete();
-        }
         exec(cmd, function () {
-            if (lintFlag && i === 0) {
-                var lint = jake.Task['lint'];
-                lint.addListener('complete', function () {
-                    complete();
-                });
-                lint.invoke();
+            if (i === 0) {
```
Comment (sandersn) [inline Jakefile.js]: shouldn't need this check for the non-parallel version

Comment (vladima) [inline Jakefile.js]: sure

File: src/harness/runnerbase.ts [hunk 2]
```diff
@@ -1,5 +1,10 @@
 /// <reference path="harness.ts" />
 
+
+type TestRunnerKind = CompilerTestKind | FourslashTestKind | "project" | "rwc" | "test262";
+type CompilerTestKind = "conformance" | "compiler";
+type FourslashTestKind = "fourslash" | "fourslash-shims" | "fourslash-shims-pp" | "fourslash-server";
```
Comment (sandersn) [inline src/harness/runnerbase.ts]: these don't need to be in order? That's strange

**Comment by sandersn** [comment]
:+1:

File: Jakefile.js [hunk 3]
```diff
@@ -717,45 +730,69 @@ function runConsoleTests(defaultReporter, defaultSubsets) {
...
-                lint.invoke();
-            }
+            runLinter();
             finish();
         }, finish);
-    });
+        
+    }
+    else {
+        // run task to load all tests and partition then between workers
```
Comment (yuit) [inline Jakefile.js]: nit: partition them

**Comment by mhegazy** [comment]
:+1:

**Pull request merged**